tests/components/modbus/test_init.py

```diff
@@ -467,3 +474,116 @@ async def test_pymodbus_connect_fail(hass, caplog, mock_pymodbus):
+                    },
+                ],
+                CONF_SENSORS: [
+                    {
+                        CONF_INPUT_TYPE: CALL_TYPE_REGISTER_HOLDING,
+                        CONF_NAME: f"{TEST_SENSOR_NAME}_3",
+                        CONF_ADDRESS: 53,
+                        CONF_SCAN_INTERVAL: 5,
+                    },
+                    {
+                        CONF_INPUT_TYPE: CALL_TYPE_REGISTER_INPUT,
+                        CONF_NAME: f"{TEST_SENSOR_NAME}_4",
+                        CONF_ADDRESS: 54,
+                        CONF_SCAN_INTERVAL: 5,
+                    },
+                ],
+            }
+        ]
+    }
+    mock_pymodbus.read_coils.return_value = ReadResult([0x01])
+    mock_pymodbus.read_discrete_inputs.return_value = ReadResult([0x01])
+    mock_pymodbus.read_holding_registers.return_value = ReadResult([7])
+    mock_pymodbus.read_input_registers.return_value = ReadResult([7])
+    now = dt_util.utcnow()
+    with mock.patch("homeassistant.helpers.event.dt_util.utcnow", return_value=now):
+        assert await async_setup_component(hass, DOMAIN, config) is True
+        await hass.async_block_till_done()
+
+    now = now + timedelta(seconds=10)
+    with mock.patch("homeassistant.helpers.event.dt_util.utcnow", return_value=now):
+        async_fire_time_changed(hass, now)
+        await hass.async_block_till_done()
+
+    # Check states
+    entity_id = f"{BINARY_SENSOR_DOMAIN}.{TEST_SENSOR_NAME}_1"
+    assert hass.states.get(entity_id).state == STATE_UNAVAILABLE
+    entity_id = f"{BINARY_SENSOR_DOMAIN}.{TEST_SENSOR_NAME}_2"
+    assert hass.states.get(entity_id).state == STATE_UNAVAILABLE
+    entity_id = f"{SENSOR_DOMAIN}.{TEST_SENSOR_NAME}_3"
+    assert hass.states.get(entity_id).state == STATE_UNAVAILABLE
+    entity_id = f"{SENSOR_DOMAIN}.{TEST_SENSOR_NAME}_4"
+    assert hass.states.get(entity_id).state == STATE_UNAVAILABLE
+
+    mock_pymodbus.reset_mock()
+    data = {
+        ATTR_HUB: TEST_MODBUS_NAME,
+        ATTR_UNIT: 17,
+        ATTR_ADDRESS: 16,
+        ATTR_STATE: False,
+    }
+    await hass.services.async_call(DOMAIN, SERVICE_WRITE_COIL, data, blocking=True)
+    assert not mock_pymodbus.write_coil.called
+    await hass.services.async_call(DOMAIN, SERVICE_WRITE_COIL, data, blocking=True)
+    assert not mock_pymodbus.write_coil.called
+    data[ATTR_STATE] = [True, False, True]
+    await hass.services.async_call(DOMAIN, SERVICE_WRITE_COIL, data, blocking=True)
+    assert not mock_pymodbus.write_coils.called
+
+    del data[ATTR_STATE]
+    data[ATTR_VALUE] = 15
+    await hass.services.async_call(DOMAIN, SERVICE_WRITE_REGISTER, data, blocking=True)
+    assert not mock_pymodbus.write_register.called
+    data[ATTR_VALUE] = [1, 2, 3]
+    await hass.services.async_call(DOMAIN, SERVICE_WRITE_REGISTER, data, blocking=True)
+    assert not mock_pymodbus.write_registers.called
+
+    # 2 times fire_changed is needed to secure "normal" update is called.
```

Comment:
I don't follow why we need to wait 2 x 30 seconds. Hasn't the 15 second delay already expired within the first 30 seconds?

Reply:
It might very well be that the interval can be set a lot shorter. What I saw in test was that (for a reason I do not understand) 2 fire_changed time was needed. I suppose it has something to do with how the event queue are created and emptied.

While I do the change above, I will play with smaller delays. However it does not make the test faster :-)

Reply:
Reduced to 6seconds and 10seconds, scraping it lower might be possible, but I experienced that repeated run of the test would sometime break.

Reply:
I think I understand what is happening now.

Time doesn't actually pass in the test. We just fire the fake time changed event. The first time changed event where time is more than the delay will call the callback to reset the delay. Then another time changed event is required to actually update the state of the entities.

Does that make sense? That would mean that the new times don't update the state reliably but depends on the order of tasks.

Reply:
your understanding is correct, I suspect it depends on the order of the events (call_later and the scheduled update, but admitted I did not debug the core event handler.

I have seen this earlier in other tests too, if you have 2 listeners with events in the same timeslot the sequence can vary.